### PR TITLE
Fix #497. Added feature Mutating Updates

### LINKED ISSUE #497: Mutating updates: update foo set a=a+123
_[Migrated from Assembla ticket [298](https://www.assembla.com/spaces/typesafe-slick/tickets/298) - reported by @cvogt on 2013-07-19 23:38:51]_

lifted embedding could support this

https://groups.google.com/forum/#!topic/scalaquery/TbrIVuyqXzE

### PR DESCRIPTION
**Pull request by radsaggi**
~~There is a conflict of approach in implementing this feature. The current branch proceeds with a minimalist change policy. This approach seems to be more reasonable since it states that the compilation cannot proceed without the select part of the mutating update and hence the `mutatingUpdateCompiler` would not be completely defined unless we have the Select AST.~~

~~There is some more work in [tmp/mutating-updates](https://github.com/radsaggi/slick/tree/tmp/mutating-updates) that employs a different policy that could be helpful in other scenarios and provides expandability.~~

Current Issues:
~~1. (SELECT col1 FROM table2 ...), col2 = (SELECT col2 FROM table2 ...) ...`~~
2. Find a way to get the current query in the `UpdateActionExtensionMethodsImpl` view so that we can do something like `def update(f: Query => Rep)`
This is to implement `UPDATE table SET a = a + 1` as `table.filter(_.id === 1).map(_.a).update(_.a + 1)`

### CODE REVIEW

**Comment by radsaggi** [comment]
We have an almost complete implementation for Mutating Updates here. The approach has undergone a sea-change.

**Concise description**
We have added `update` methods to a `Query` and it produces an `UpdateQuery` that wraps an `Update` node. `UpdateQuery` is implicitly converted to a `ProfileAction` as required. Compilation of this node proceeds in three major steps using a modified compiler that adds a `ResolveUpdate` phase. 

1. The `ResolveUpdate` phase converts an `Update` node to an `UpdateBind` by breaking the columns and their matching values into an individual `Update` node. Hence the `UpdateBind` has an AST that conforms to `UpdateBind(gen, setNode, StructNode(updates), hasSubSelects)`. 
2. This is then transformed again in the `MergeToComprehensions` phase and a `Comprehension` with a new field `updateValues: ConstArray[Node]` is generated.
3. This Comprehension is interpreted and compiled into a SQL statement by the existing `CodeGen` with some tweaks for handling the `updateValues`.

**What remains**
Figure out how a Table's `ProvenShape` can be used to accept queries of the following nature:
```scala
case class Data(a: Int, b: String)

class T1(tag: Tag) extends Table[Data](tag, "T_mutating1") {
  def a = column[Int]("A")
  def b = column[String]("B")
  def * = (a, b).mapTo[Data]
}
val ts1 = TableQuery[T1]

ts1.filter(_.a === 11).update(d => Data(d.a + 200, "newData"))
```
Although the same will work as following with existing code:
```scala
ts1.filter(_.a === 11).map(d => (d.a, d.b)).update(d => (d.a + 200, "newData"))
```

Check Compiled Update queries and how they might be affected

@szeiger @cvogt: Please review!

**Comment by radsaggi** [comment]
Apparently the tests are failing at doc generation. I would update the same once we are through the review process and have finalized the API.

File: slick/src/main/scala/slick/sql/SqlCapabilities.scala [hunk 1]
```diff
@@ -16,6 +16,8 @@ object SqlCapabilities {
   val sequenceMax = Capability("sql.sequenceMax")
   /** Supports min value for sequences */
   val sequenceMin = Capability("sql.sequenceMin")
+  /** supports interpreting strings as VARCHAR. Some databases eg. HSQL do not do this. */
+  val stringIsVarchar = Capability("sql.stringIsVarchar")
```
Comment (cvogt) [inline slick/src/main/scala/slick/sql/SqlCapabilities.scala]: is this really an HSQL limitation or a design choice slick made?

File: slick/src/main/scala/slick/util/ConstArray.scala [hunk 2]
```diff
@@ -433,6 +433,16 @@ object ConstArray {
...
+  def tabulate[T](n: Int)(f: Int => T): ConstArray[T] = {
+    val a = new Array[Any](n)
+    var i = 0
+    while(i < n) {
+      a(i) = f(i)
+      i += 1
+    }
+    new ConstArray[T](a)
+  }
+
```
Comment (cvogt) [inline slick/src/main/scala/slick/util/ConstArray.scala]: this doesn't seem to be used

File: slick/src/main/scala/slick/sql/SqlCapabilities.scala (hunk 1 above)
Comment (radsaggi) [inline slick/src/main/scala/slick/sql/SqlCapabilities.scala]: I read about this [here](https://github.com/slick/slick/blob/master/slick/src/main/scala/slick/jdbc/HsqldbProfile.scala#L76).
Hence I am using this capability. It might be useful to make the description more verbose. Should I do that?

File: slick/src/main/scala/slick/util/ConstArray.scala (hunk 2 above)
Comment (radsaggi) [inline slick/src/main/scala/slick/util/ConstArray.scala]: It's a good feature to have nonetheless. Maybe I should move it another PR?

**Comment by szeiger** [comment]
Haven't looked at the implementation details yet, but I think

    ts1.filter(_.a === 11).update(d => Data(d.a + 200, "newData"))

is unattainable in the Lifted Embedding. `Data` is a non-lifted constructor. I can't think of any implicits trickery you could use to make this work in a way that is as generic as we need. You'll have to stay in the lifted domain, which is `Tuple2` in this case. It should be possible to use both, monomorphic and polymorphic record types. Since it already works for Tuple2 and the machinery is the same for all of those, that should already work (modulo minor bug fixes perhaps).

**Comment by szeiger** [comment]
In practice it could look like this:

```
case class Data(a: Int, b: String)
case class LiftedData(a: Rep[Int], b: Rep[String])
implicit object dataShape extends CaseClassShape(LiftedData.tupled, Data.tupled)

class T1(tag: Tag) extends Table[Data](tag, "T_mutating1") {
  def a = column[Int]("A")
  def b = column[String]("B")
  def * = LiftedData(a, b)
}
val ts1 = TableQuery[T1]

ts1.filter(_.a === 11).update(d => LiftedData(d.a + 200, "newData"))
```

File: slick/src/main/scala/slick/sql/SqlCapabilities.scala (hunk 1 above)
Comment (szeiger) [inline slick/src/main/scala/slick/sql/SqlCapabilities.scala]: I don't think it belongs in SqlCapabilities. If it's for testing only, put it into TestDB.capabilities. The distinction is handled internally in HsqldbDriver anyway, without looking at this flag.

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMapperTest.scala [hunk 3]
```diff
@@ -76,6 +76,83 @@ class JdbcMapperTest extends AsyncTest[JdbcTestDB] {
...
+//    val q4: SqlDBIO[Int] = updateQ14.update3(d => (d.a + 200, "tuple"))
+    val q5: SqlDBIO[Int] = updateQ15.update3(d => d ++ "T1")
+    val q6: SqlDBIO[Int] = updateQ16.update3[(Rep[String], Rep[Int])]{ case (b, a) => (b ++ "T1", a + 100) }
+    val q7: SqlDBIO[Int] = updateQ14.update(Data(1, "1"))
+    val q8: SqlDBIO[Int] = updateQ15.update3("T0")
+    val q9: SqlDBIO[Int] = updateQ16.update3("T0", 57)
+
+    // Exclude this so that string casting(eg in hsqldb) may not affect the test against constant sql strings
+    ifCapU(scap.stringIsVarchar) {
+      q1.statements.head shouldBe
```
Comment (szeiger) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMapperTest.scala]: Checking generated statements is very brittle. It's usually better to *run* tests and check the results. (Of course, this requires coming up with a good set of test data)

File: slick/src/main/scala/slick/ast/Comprehension.scala [hunk 4]
```diff
@@ -6,17 +6,19 @@ import slick.util.ConstArray
 
 /** A SQL comprehension */
 final case class Comprehension(sym: TermSymbol, from: Node, select: Node, where: Option[Node] = None,
+                               updateValues: ConstArray[Node] = ConstArray.empty,
```
Comment (szeiger) [inline slick/src/main/scala/slick/ast/Comprehension.scala]: Why is `updateValues` an array? `select` is a single node at this point. Since Comprehensions only show up after the flattening phases, it should always be a `StructNode` or `ProductNode` (depending on the phase). Shouldn't `updateValues` be the same?

File: slick/src/main/scala/slick/ast/Comprehension.scala [hunk 5]
```diff
@@ -79,6 +85,7 @@ final case class Comprehension(sym: TermSymbol, from: Node, select: Node, where:
         from = f2,
         select = s2,
         where = w2.orElse(where),
+        updateValues = if (u2 eq updateValues) updateValues else u2,
```
Comment (szeiger) [inline slick/src/main/scala/slick/ast/Comprehension.scala]: unnecessary check

File: slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMapperTest.scala (hunk 3 above)
Comment (radsaggi) [inline slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JdbcMapperTest.scala]: I guess I put this in more as a demonstration of what the users could expect the statements to be like than as a test. The queries are being run and tested as well.

File: slick/src/main/scala/slick/sql/SqlCapabilities.scala (hunk 1 above)
Comment (radsaggi) [inline slick/src/main/scala/slick/sql/SqlCapabilities.scala]: Done!

**Comment by SethTisue** [comment]
closing for inactivity
